main_test.go

```diff
@@ -56,32 +56,87 @@ func BenchmarkKubeStateMetrics(t *testing.B) {
 	builder.WithKubeClient(kubeClient)
 	builder.WithNamespaces(options.DefaultNamespaces)
 
-	collectors := builder.Build()
+	// This test is not suitable to be compared in terms of time, as it includes
+	// a one second wait. Use for memory allocation comparisons, profiling, ...
+	t.Run("GenerateMetrics", func(b *testing.B) {
+		collectors = builder.Build()
 
-	handler := metricHandler{collectors, false}
+		// Wait for caches to fill
+		time.Sleep(time.Second)
+	})
 
+	handler := metricHandler{collectors, false}
 	req := httptest.NewRequest("GET", "http://localhost:8080/metrics", nil)
 
-	// Wait for informers to sync
+	t.Run("MakeRequests", func(b *testing.B) {
+		for i := 0; i < requestCount; i++ {
+			w := httptest.NewRecorder()
+			handler.ServeHTTP(w, req)
+
+			resp := w.Result()
+			if resp.StatusCode != 200 {
```

Comment:
In general, you are just checking that the status is OK, but not that the data that is being served is present. I guess what I am suggesting is that maybe you should check that the metrics you are benchmarking are actually there.

Reply:
Agreed. Maybe without checking the "right" content, we can just use the `BenchmarkTest.SetBytes` functionality of the benchmark to set the amount of bytes processed, if things are completely broken then the bytes amount should be off as well, further validation would be done by e2e tests.

Reply:
I am a bit undecided here. On the one hand I would like  to separate concerns, thereby not do correctness testing in a benchmark test, on the other hand I see your point and would also like to test correct responses, especially under load.

In regards to using  `BenchmarkTest.SetBytes`: Yes, that would be an option but requires a user to monitor this metric, whereas we could also entirely fail if the result is wrong.

I will look into stopping the timer and then properly checking the result.

Reply:
We could compare the output to a [golden record](https://www.youtube.com/watch?v=yszygk1cpEc) but that would need to be touched every time we change a metric.

Thinking about this some more I would prefer to keep it as it is with the 200 check, as there is the correctness check further below (`TestFullScrapeCycle`). What do you think @LiliC @brancz?

Reply:
Yes the actual correctness should not be tested in the benchmark I agree, real e2e tests are best for that, our unit tests are already very close to the golden record type testing, and we have e2e tests that check that every collector does expose metrics.

I still think setting bytes is a useful thing, and possibly could show users early that things are completely wrong, but not necessarily a complete check. Nevertheless showing how much Mb/s we can do at exposition is interesting to know.

Reply:
> I am a bit undecided here. On the one hand I would like to separate concerns, thereby not do correctness testing in a benchmark test, on the other hand I see your point and would also like to test correct responses, especially under load.

Yes this was my thought process behind my initial comment suggestion, was not sure either.

> Thinking about this some more I would prefer to keep it as it is with the 200 check, as there is the correctness check further below (TestFullScrapeCycle).

Agreed, it makes perfect sense. 👍